haystack/document_stores/opensearch.py

```diff
@@ -598,47 +506,220 @@ def _create_document_index(self, index_name: str, headers: Optional[Dict[str, st
+                f"'{embedding_field_knn_engine}', but knn_engine was set to '{self.knn_engine}'. "
+                f"To switch knn_engine to '{self.knn_engine}' consider one of these options: "
+                f" - Clone the embedding field in the same index, e.g. `clone_embedding_field(knn_engine='{self.knn_engine}', ...)`. "
+                f" - Create a new index by selecting a different index name, e.g. `index='my_new_{self.knn_engine}_index'`. "
+                f" - Overwrite the existing index by setting `recreate_index=True`. Note that all existing data will be lost!"
+            )
+
+        # Check method params according to requested index_type
         if self.index_type == "flat":
-            # use default parameters from https://opensearch.org/docs/1.2/search-plugins/knn/knn-index/
-            # we need to set them explicitly as aws managed instances starting from version 1.2 do not support empty parameters
-            method["parameters"] = {"ef_construction": 512, "m": 16}
-        elif self.index_type == "hnsw":
-            method["parameters"] = {"ef_construction": 80, "m": 64}
-            # for nmslib this is a global index setting
+            self._assert_embedding_param(
+                name="method.name", actual=embedding_field_method_name, expected="hnsw", index_id=index_id
+            )
+            self._assert_embedding_param(
+                name="ef_construction", actual=embedding_field_ef_construction, expected=512, index_id=index_id
+            )
+            self._assert_embedding_param(name="m", actual=embedding_field_m, expected=16, index_id=index_id)
             if self.knn_engine == "faiss":
-                method["parameters"]["ef_search"] = 20
-        else:
-            logger.error("Please set index_type to either 'flat' or 'hnsw'")
+                self._assert_embedding_param(
+                    name="ef_search", actual=embedding_field_ef_search, expected=512, index_id=index_id
+                )
+        if self.index_type == "hnsw":
+            self._assert_embedding_param(
+                name="method.name", actual=embedding_field_method_name, expected="hnsw", index_id=index_id
+            )
+            self._assert_embedding_param(
+                name="ef_construction", actual=embedding_field_ef_construction, expected=80, index_id=index_id
+            )
+            self._assert_embedding_param(name="m", actual=embedding_field_m, expected=64, index_id=index_id)
+            if self.knn_engine == "faiss":
+                self._assert_embedding_param(
+                    name="ef_search", actual=embedding_field_ef_search, expected=20, index_id=index_id
+                )
+
+    def _assert_embedding_param(self, name: str, actual: Any, expected: Any, index_id: str) -> None:
+        if actual != expected:
+            message = (
+                f"Existing embedding field '{self.embedding_field}' of OpenSearch index '{index_id}' has {name} value '{actual}', "
```

Comment:
Can we make it easier to understand?

The index_type {index type} needs {expected} as value. Currently, the value set for it is {value}.